### PR TITLE
Updating qnn.KerasLayer to work in tape mode

### PR DESCRIPTION
**Pull request by trbromley**
**Context:** The `qnn.KerasLayer` class needs modification to work in tape mode, due to the removal of QNode signature introspection.

**Description of changes:**

* The `KerasLayer` class requires different logic for signature validation and qnode evaluation in tape mode, so the new methods `_signature_validation_tape_mode` and `_evaluate_qnode_tape_mode`, to be called instead if in tape mode. This resulted in some 'rearrangement' of the class, but the underlying changes are minimal.

* The `get_circuit()` fixture is copied into `test_keras_layer.py`, since it needs to be modified for tape mode, _and_ the Torch class hasn't been ported yet (at least in `master`).

* Some of the tests would previously use the `get_circuit` fixture, but monkeypatch the created valid circuit to be non-valid, to test validation. Due to deviations between tape mode and non-tape mode, it is simpler to simply create a small, invalid test QNode within the test itself. This also makes the test more realistic, as sometimes monkeypatching would not replicate the user experience.

* **Support for backprop diff_method has been provided**

**Benefits & drawbacks:**

Similar to in https://github.com/PennyLaneAI/pennylane/pull/865.

### CODE REVIEW

File: pennylane/qnn/keras.py [hunk 1]
```diff
@@ -270,24 +306,40 @@ def call(self, inputs):
...
+    def _evaluate_qnode_tape_mode(self, x):
+        """Evaluates a tape-mode QNode for a single input datapoint.
+
+        Args:
+            x (tensor): the datapoint
+
+        Returns:
+            tensor: output datapoint
+        """
+        kwargs = {**{self.input_arg: x}, **{k: 1.0 * w for k, w in self.qnode_weights.items()}}
```
Comment (trbromley) [inline pennylane/qnn/keras.py:340]: Note the 1.0 here, which converts a `Variable` into a `Tensor`, allowing us to do things like `qml.Rot(*weight)`.

File: pennylane/tape/qnode.py [hunk 2]
```diff
@@ -150,6 +150,7 @@ def __init__(
         self.func = func
         self.device = device
         self.qtape = None
+        self.qfunc_output = None
```
Comment (trbromley) [inline pennylane/tape/qnode.py:153]: Everything here is a direct copy of https://github.com/PennyLaneAI/pennylane/pull/865, so please defer to that for code review.

File: pennylane/qnn/keras.py [hunk 3]
```diff
@@ -198,7 +199,32 @@ def __init__(
...
+            weight: (tuple(size) if isinstance(size, Iterable) else (size,) if size > 1 else ())
+            for weight, size in weight_shapes.items()
+        }
+
+        if qml.tape_mode_active():
+            self._signature_validation_tape_mode(qnode, weight_shapes)
+            self.qnode = qnode
+
+            dtype = tf.float32 if tf.keras.backend.floatx() == tf.float32 else tf.float64
+            self.qnode.to_tf(dtype=dtype)
```
Comment (trbromley) [inline pennylane/qnn/keras.py:212]: I have no idea why, but `self.qnode.to_tf(dtype=tf.keras.backend.floatx())` causes an error when calculating the gradient. The error looks like we're setting the `dtype` to be `tf.float64` when it's expecting `tf.float32`. However, I print out `tf.keras.backend.floatx()` and it always is `tf.float32` :shrug:

File: tests/qnn/test_keras.py [hunk 4]
```diff
@@ -138,6 +177,52 @@ class FuncPatch:
...
+            qml.RX(kwargs["w7"], wires=4 % n_qubits)
+            return [qml.expval(qml.PauliZ(i)) for i in range(output_dim)]
+
+        layer = KerasLayer(c, w, output_dim=output_dim)
+        x = tf.ones((2, n_qubits))
+
+        layer_out = layer(x)
+        circ_weights = layer.qnode_weights.copy()
+        circ_weights["w4"] = 1.0 * circ_weights["w4"]
+        circ_weights["w6"] = 1.0 * circ_weights["w6"]
```
Comment (trbromley) [inline tests/qnn/test_keras.py:221]: This is so that we don't have problems with `Variable` slicing in TF (also used a few more times below)

File: tests/qnn/test_keras.py [hunk 5]
```diff
@@ -335,16 +470,49 @@ def test_gradients(self, get_circuit, output_dim, n_qubits):
...
+
+        with tf.GradientTape() as tape:
+            out = tf.reduce_sum(qlayer(inputs))
+
+        spy = mocker.spy(qml.tape.tapes.QubitParamShiftTape, "jacobian")
+
+        grad = tape.gradient(out, qlayer.trainable_weights)
+        assert grad is not None
+        spy.assert_not_called()
+
```
Comment (trbromley) [inline tests/qnn/test_keras.py:513]: This is another part that's different from the other PR.

File: pennylane/qnn/keras.py (hunk 1 above)
Comment (co9olguy) [reply pennylane/qnn/keras.py:344]: sneaky

File: tests/qnn/test_keras.py [hunk 6]
```diff
@@ -101,28 +142,26 @@ def test_weight_shape_unspecified(self, get_circuit, output_dim):
...
-                KerasLayer(c, w, output_dim)
+        with pytest.raises(TypeError, match="Cannot have a variable number of positional"):
+            KerasLayer(circuit, weight_shapes, output_dim=1)
 
     @pytest.mark.parametrize("n_qubits, output_dim", indices_up_to(1))
     def test_var_keyword(self, get_circuit, monkeypatch, output_dim):
         """Test if a TypeError is raised when instantiated with a variable number of keyword
         arguments"""
+        if qml.tape_mode_active():
+            pytest.skip("This functionality is supported in tape mode.")
```
Comment (co9olguy) [inline tests/qnn/test_keras.py:163]: Just wanted to confirm this is as intended, i.e., no typos? (just looks a bit unusual :laughing:)

File: tests/qnn/test_keras.py [hunk 7]
```diff
@@ -180,6 +268,53 @@ def c_dummy(inputs, w1, w2, w3, w4, w5, w6, w7, w8=None):
...
+            parameters, w5, are specified as non-trainable."""
+            qml.templates.AngleEmbedding(inputs, wires=list(range(n_qubits)))
+            qml.templates.StronglyEntanglingLayers(w1, wires=list(range(n_qubits)))
+            qml.RX(w2[0], wires=0 % n_qubits)
+            qml.RX(w3, wires=1 % n_qubits)
+            qml.Rot(*w4, wires=2 % n_qubits)
+            qml.templates.StronglyEntanglingLayers(w5, wires=list(range(n_qubits)))
+            qml.Rot(*w6, wires=3 % n_qubits)
+            qml.RX(w7, wires=4 % n_qubits)
+            return [qml.expval(qml.PauliZ(i)) for i in range(output_dim)]
```
Comment (co9olguy) [inline tests/qnn/test_keras.py:305]: How come this fixture is reproduced directly here? :thinking:

File: tests/qnn/test_keras.py [hunk 8]
```diff
@@ -180,6 +268,53 @@ def c_dummy(inputs, w1, w2, w3, w4, w5, w6, w7, w8=None):
...
+
+        layer = KerasLayer(c, w, output_dim=output_dim)
+        x = tf.ones((2, n_qubits))
+
+        layer_out = layer(x)
+        circ_weights = layer.qnode_weights.copy()
+        circ_weights["w4"] = 1.0 * circ_weights["w4"]
+        circ_weights["w6"] = 1.0 * circ_weights["w6"]
```
Comment (co9olguy) [inline tests/qnn/test_keras.py:313]: Might be good to leave a comment in the test to indicate why you're doing this

File: tests/qnn/test_keras.py [hunk 9]
```diff
@@ -335,16 +470,49 @@ def test_gradients(self, get_circuit, output_dim, n_qubits):
 
         g_layer = tape.gradient(out_layer, layer.trainable_variables)
 
+        circuit_weights = layer.trainable_variables.copy()
+        circuit_weights[3] = 1.0 * circuit_weights[3]
+        circuit_weights[5] = 1.0 * circuit_weights[5]
```
Comment (co9olguy) [inline tests/qnn/test_keras.py:475]: ditto

File: tests/qnn/test_keras.py (hunk 5 above)
Comment (co9olguy) [reply tests/qnn/test_keras.py:513]: I looked at the wrong one first it seems :frowning_face:

File: tests/qnn/test_keras.py [hunk 10]
```diff
@@ -335,16 +470,49 @@ def test_gradients(self, get_circuit, output_dim, n_qubits):
...
+
+        spy = mocker.spy(qml.tape.tapes.QubitParamShiftTape, "jacobian")
+
+        grad = tape.gradient(out, qlayer.trainable_weights)
+        assert grad is not None
+        spy.assert_not_called()
+
 
-@pytest.mark.parametrize("interface", qml.qnodes.decorator.ALLOWED_INTERFACES)
+@pytest.mark.parametrize("interface", ["autograd", "torch", "tf"])
```
Comment (co9olguy) [inline tests/qnn/test_keras.py:515]: How come this was updated? Was the old way not sufficient? Not suited to testing both old and new modes?

File: tests/qnn/test_keras.py [hunk 11]
```diff
@@ -335,16 +470,49 @@ def test_gradients(self, get_circuit, output_dim, n_qubits):
...
         g_circuit = tape.gradient(out_circuit, layer.trainable_variables)
 
         for i in range(len(out_layer)):
             assert np.allclose(g_layer[i], g_circuit[i])
 
+    @pytest.mark.parametrize("n_qubits, output_dim", indices_up_to(1))
+    def test_backprop_gradients(self, mocker):
+        """Test if KerasLayer is compatible with the backprop diff method."""
+        if not qml.tape_mode_active():
+            pytest.skip("This functionality is only supported in tape mode.")
```
Comment (co9olguy) [inline tests/qnn/test_keras.py:489]: Awesome

**Review by co9olguy — APPROVED** [review]
The subtleties behind some of the signature inspection stuff is not my expertise, so would be good to get another pair of eyes to look over

File: tests/qnn/test_keras.py (hunk 10 above)
Comment (josh146) [reply tests/qnn/test_keras.py:515]: Using `interface="numpy"` has been deprecated for a while; ever since we introduced TF/Torch support. The supported name is now `interface="autograd"`. In the new tape-mode, attempting to do `interface="numpy"` will now raise an error.

We could uses the new location of the allowed interface names:
```python
@pytest.mark.parametrize("interface", qml.tape.QNode.INTERFACE_MAP)
```

I always tend to prefer being explicit in the tests, though. Better that parametrized data is test-specific, rather than dependent on logic/implementation details from the codebase. If, at any point a new interface is added, ideally we should be notified to update tests due to reduced coverage.

**Review by josh146 — APPROVED** [review]
Really really awesome that `KerasLayer` now supports backprop 💪 

I only had a minor comment regarding the casting from Variables to Tensors - is this something we should leave to the user, rather than attempting to do implicitly behind the scenes?

File: tests/qnn/test_keras.py (hunk 8 above)
Comment (trbromley) [reply tests/qnn/test_keras.py]: Thanks! I followed Josh's suggestion of using `tf.convert_to_tensor()` and also added a comment that this is for slicing.

File: tests/qnn/test_keras.py (hunk 11 above)
Comment (trbromley) [reply tests/qnn/test_keras.py:493]: Nice, added to changelog!
